orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -105,52 +135,60 @@ def _forward(self, *inputs, **kwargs):
-                device_from_module = _utils.get_device_from_module(
-                    self._original_module)
-                if not self._device or self._device != device_from_module:
-                    self._device = device_from_module
-                    if not self._device:
-                        raise RuntimeError(
-                            'A device must be specified in the model or data!')
+            # Flag to indicate whether the gradient_graph needs to be built
+            # It should be built every time the module gradient graph builder is initialized
+            build_gradient_graph = \
                 self._get_inference_graph_and_init_gradient_graph_builder(
                     *inputs, **kwargs)
 
-            # Flag to indicate whether the gradient_graph needs to be built
-            build_gradient_graph = self._current_input_shape is None
             _, _, input_names_require_grad, new_input_shape = \
                 _ortmodule_io.parse_inputs_for_onnx_export(
-                    self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
-            initializer_names_to_train_set_user_model = {name for name, param in
-                self._flattened_output_module.named_parameters() if param.requires_grad}
-            initializer_names_to_train_set_onnx_graph = set(self._onnx_graphs_info.initializer_names_to_train) \
-                if self._onnx_graphs_info else None
-            # If inputs requiring gradient change from forward to the next, the module_gradient_graph_builder
-            # needs to be reinitialized so it can compute the backward output for the new inputs that require_grad
-            if input_names_require_grad != self._input_names_require_grad or \
-                initializer_names_to_train_set_user_model != initializer_names_to_train_set_onnx_graph:
-                self._input_names_require_grad = input_names_require_grad
-                self._initialize_module_gradient_graph_builder()
-                # Trigger the rebuilding of the gradient graph
-                build_gradient_graph = True
-
+                    self._original_module_parameters, self._exported_onnx_training_model if self._is_training() else self._exported_onnx_inference_model, *inputs, **kwargs)
+            # Reinitialize gradient graph builder if the mode is training mode,
+            # and the inputs or initializers requiring gradient have changed.
+            build_gradient_graph = build_gradient_graph or \
+                self._reinitialize_module_gradient_graph_builder(input_names_require_grad)
+
+            # Build the gradient graph if the model is in train mode,
+            # or build the inference graph if model is in eval mode
             if build_gradient_graph:
                 self._current_input_shape = new_input_shape
                 self._build_training_graph()
-                self._create_training_session()
+                self._build_inference_graph()
```

Comment:
I think it'd be a little more readable to move the `if self._is_training()` test outside these methods (rather than making them a no-op in the "wrong" mode):

```suggestion
                if self._is_training():
                  self._build_training_graph()
                else:
                  self._build_inference_graph()
```

Reply:
Reformatted the code, so this check does not come into play as much now.